### PR TITLE
✨ Write the 2018 retrospective blog post

### PR DESCRIPTION
**Pull request by tomhoule**
[rendered](https://github.com/store2be/store2be.github.io/blob/b11a0facbbd1e67c13a9379a8a7aed4f015ea536/_posts/2019-01-30-year-of-the-dog.md)

### CODE REVIEW

File: _posts/2019-01-30-year-of-the-dog.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+![store2be developer, early 2018]({{ site.url }}/images/2018_retrospective/grumpy_pallas_cat_c_tambako_the_jaguar_flickr.jpg)
+_store2be developer, early 2018 (photo © Tambako The Jaguar, flickr)_
+
+## Frontend - the React apps
+
+We introduced Typescript in our React apps at the end of 2017, and we are only
+now getting close to completely typed frontend apps. Before Typescript, we had
+suffered from the churn in our JS architecture as we experimented with more
+scaleable approaches to state management and asynchronous actions - we had our
```
Comment (brainlessdeveloper) [inline _posts/2019-01-30-year-of-the-dog.md]: `scalable`

File: _posts/2019-01-30-year-of-the-dog.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+stateful apps without introducing type errors which translated directly to
+crashes for our users. There are enough Typescript success stories on the
+internet, so I will not elaborate too much on this.
+
+Good tooling is something we appreciate, so in addition to Typescript we
+adopted [prettier](https://prettier.io/),
+[TSLint](https://github.com/palantir/tslint/) (in addition to ESLint),
+[jest](https://jestjs.io/) (replacing jasmine), and [snapshot
+testing](https://jestjs.io/docs/en/snapshot-testing.html). All of these lifted
+a lot of mental burden: no more worrying about or manually fiddling with code
```
Comment (brainlessdeveloper) [inline _posts/2019-01-30-year-of-the-dog.md]: `All of these helped lift mental burden`, I think "burden" can't be used as uncountable.

File: _posts/2019-01-30-year-of-the-dog.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+formatting, automatically enforced conventions, more certainty that the UI does
+not change by surprise, etc. Linting, testing and a lot more is enforced in CI so
+we can focus on the hard problems during code reviews.
+
+We liked the experience so much we implemented a backend service for email templating
+in typescript and [wrote about
+it](https://tech.store2be.com/email/sendwithus/mjml/typescript/react/2018/06/14/email-templates-at-store2be-and-gdpr/).
+
+Another notable and much appreciated addition to our toolbox was
```
Comment (brainlessdeveloper) [inline _posts/2019-01-30-year-of-the-dog.md]: `much-appreciated`

File: _posts/2019-01-30-year-of-the-dog.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+[react-storybook](https://github.com/storybooks/storybook) for our shared UI
+components.
+
+## An unexpected journey - GraphQL
+
+In day to day frontend work however, architectural issues were still making our lives hard:
```
Comment (brainlessdeveloper) [inline _posts/2019-01-30-year-of-the-dog.md]: `work, however,`

File: _posts/2019-01-30-year-of-the-dog.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+1. Our RESTful API structure had not evolved much: most of our endpoints were
+   mapping directly to database table, with little additional logic.  With more
```
Comment (brainlessdeveloper) [inline _posts/2019-01-30-year-of-the-dog.md]: `to a database table` or `to database tables`

File: _posts/2019-01-30-year-of-the-dog.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+   resources and more complex business logic, that meant **a lot of business
+   logic had to be implemented in the frontend** and merely validated by the
+   API.
+2. **State management and synchronization in the React apps was getting too
+   complex** - some operations and queries required several network calls and
+   dispatching multiple redux actions. Some critical operations were not
+   atomic, when they should have been.
```
Comment (brainlessdeveloper) [inline _posts/2019-01-30-year-of-the-dog.md]: `were not atomic when they should have been` (comma)

File: _posts/2019-01-30-year-of-the-dog.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+3. **Onboarding new developers to our redux-saga based abstractions** and all
+   the related conventions **proved challenging**, as documentation was not as
+   comprehensive as it should have been and old conventions coexisted with new
```
Comment (brainlessdeveloper) [inline _posts/2019-01-30-year-of-the-dog.md]: `been, and`

File: _posts/2019-01-30-year-of-the-dog.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+  had to come up with a test-friendly code structure.
+- **Limiting access and visibility**. Nothing particularly challenging here,
+  but this also needs to be learned.
+- **Proper use of more "advanced" features like interfaces**
+- **Avoiding the n+1 query problem** - there is one well-supported,
+  well-documented and straightforward solution,
+  [graphql-batch](https://github.com/Shopify/graphql-batch) (by Shopify), but
+  this is one more thing to learn and implement.
+
+It took us a few months to arrive to satisfactory abstractions. Resources like
```
Comment (brainlessdeveloper) [inline _posts/2019-01-30-year-of-the-dog.md]: ~arrive to~ -> `achieve` I don't think `arrive to` works

File: _posts/2019-01-30-year-of-the-dog.md [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+   overall amount of state management code has decreased. State is (or
+   appears) also more localized to where it belongs in the component tree. We
+   found this easier to reason about.
+3. The simplicity, but also the adherence to a well-adopted and documented
+   standard for data fetching made **onboarding easier**. There are a lot of
+   good GraphQL resources for frontend developers. In our experience picking up
+   the technology was easy for everyone.
+4. The tooling is amazing, it lets us **leverage Typescript's strengths by
+   generating precise types for our API data**. We use apollo-client and
+   apollo-codegen, they are well-documented, worked fine and the caching layer
```
Comment (peterfication) [inline _posts/2019-01-30-year-of-the-dog.md]: You could add links to apollo here.

File: _posts/2019-01-30-year-of-the-dog.md [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+   from the redux-saga generators, we felt that we were **not realizing
+   Typescript's potential**.
+5. **Documentation was maintained manually as a Swagger specification**. This
+   was also a lot of error-prone, manual work.
+
+Faced with these problems, we decided to try and address them by progressively
+introducing a new GraphQL API, in the same Rails codebase as the existing
+RESTful API. Here is what we observed:
+
+1. The additional layer gave us more room and structure to **move business
```
Comment (peterfication) [inline _posts/2019-01-30-year-of-the-dog.md]: Maybe we can add something that our business logic seems more structured now as well. => We took the time to carefully think about a consistent structure of the backend code.

File: _posts/2019-01-30-year-of-the-dog.md [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+   apollo-codegen, they are well-documented, worked fine and the caching layer
+   is the cherry on top.
+5. **Our API documentation is now in much better shape**, because most of it is
+   automated, keeping it in sync with the implementation is trivial, and
+   [GraphiQL][graphiql] offers a good interface to explore it.
+
+It took some time to have a viable minimum implementation, but once this was
+deployed, it became clear very quickly that we wanted to go in that direction.
+
+Two more pain-points that we did not identify before migrating but are now obvious:
```
Comment (peterfication) [inline _posts/2019-01-30-year-of-the-dog.md]: I would rephrase this a little bit. For me it read like we had problems but it's actually about `pain-points that were solved`

File: _posts/2019-01-30-year-of-the-dog.md [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+
+We liked the experience so much we implemented a backend service for email templating
+in typescript and [wrote about
+it](https://tech.store2be.com/email/sendwithus/mjml/typescript/react/2018/06/14/email-templates-at-store2be-and-gdpr/).
+
+Another notable and much appreciated addition to our toolbox was
+[react-storybook](https://github.com/storybooks/storybook) for our shared UI
+components.
+
+## An unexpected journey - GraphQL
```
Comment (peterfication) [inline _posts/2019-01-30-year-of-the-dog.md]: We could change this to `An unexpected journey - GraphQL in the frontend` to make it clear, that you talk about the backend later.

File: _posts/2019-01-30-year-of-the-dog.md (hunk 10 above)
Comment (peterfication) [reply _posts/2019-01-30-year-of-the-dog.md]: ok, I didn't read the whole post before commenting ^^

File: _posts/2019-01-30-year-of-the-dog.md [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+  [graphql-batch](https://github.com/Shopify/graphql-batch) (by Shopify), but
+  this is one more thing to learn and implement.
+
+It took us a few months to arrive to satisfactory abstractions. Resources like
+the [Shopify GraphQL design
+tutorial](https://github.com/Shopify/graphql-design-tutorial) were helpful. In
+general, having someone with prior knowledge or significant interest/experience
+advocating GraphQL good practices and tools helps.
+
+Now that the abstractions are well defined, we find it easier to onboard
```
Comment (peterfication) [inline _posts/2019-01-30-year-of-the-dog.md]:
```suggestion
Now that the abstractions are now well defined, we find it easier to onboard
```

File: _posts/2019-01-30-year-of-the-dog.md [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+developers to the API, and we work more with plain ruby classes instead of
+Rails-specific methods in models and controllers.
+
+In short, the change was received enthusiastically by the whole team once the
+initial hurdles were behind us, and we are very happy with it.
+
+## Backend and infrastructure
+
+We started implementing an event log for our platform, for auditing, debugging
+and in the future decoupling tasks. It opens up a lot of perspectives and
```
Comment (peterfication) [inline _posts/2019-01-30-year-of-the-dog.md]:
```suggestion
and in the future for decoupling tasks. It opens up a lot of perspectives and
```

File: _posts/2019-01-30-year-of-the-dog.md [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+Now that the abstractions are well defined, we find it easier to onboard
+developers to the API, and we work more with plain ruby classes instead of
+Rails-specific methods in models and controllers.
+
+In short, the change was received enthusiastically by the whole team once the
+initial hurdles were behind us, and we are very happy with it.
+
+## Backend and infrastructure
+
+We started implementing an event log for our platform, for auditing, debugging
```
Comment (peterfication) [inline _posts/2019-01-30-year-of-the-dog.md]: The primary use case was the timeline feature ^^

Comment (tomhoule) [reply _posts/2019-01-30-year-of-the-dog.md]: added this :p

File: _posts/2019-01-30-year-of-the-dog.md (hunk 3 above)
Comment (indifferentalex) [reply _posts/2019-01-30-year-of-the-dog.md]: Apparently it's grammatically correct to use the hypen, but it looks weird to me ¯\_(ツ)_/¯

Comment (indifferentalex) [reply _posts/2019-01-30-year-of-the-dog.md]: But apparently it's more grammatically correct

File: _posts/2019-01-30-year-of-the-dog.md (hunk 2 above)
Comment (brainlessdeveloper) [reply _posts/2019-01-30-year-of-the-dog.md]: I think that's wrong as well: the first example should be `My family puts too much of a burden on me`.

Comment (brainlessdeveloper) [reply _posts/2019-01-30-year-of-the-dog.md]: Or `too big a burden`.

File: _posts/2019-01-30-year-of-the-dog.md (hunk 4 above)
Comment (brainlessdeveloper) [reply _posts/2019-01-30-year-of-the-dog.md]: A comma after that `however` is not a matter of style, I think.

Comment (brainlessdeveloper) [reply _posts/2019-01-30-year-of-the-dog.md]: Poor Tom though, anything is fine xD

Comment (tomhoule) [reply _posts/2019-01-30-year-of-the-dog.md]: I don't know, when I read the sentence there is no pause here.

File: _posts/2019-01-30-year-of-the-dog.md [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+Rails-specific methods in models and controllers.
+
+In short, the change was received enthusiastically by the whole team once the
+initial hurdles were behind us, and we are very happy with it.
+
+## Backend and infrastructure - events, k8s
+
+We started implementing an event log for our platform, for auditing, debugging
+and in the future for decoupling tasks. The primary feature at the moment is a
+timeline in our internal app, but it opens up a lot of perspectives and
```
Comment (indifferentalex) [inline _posts/2019-01-30-year-of-the-dog.md]: If we want to give it a bit more oomph: `which allows us to get a better overview of how various business processes work,`

File: _posts/2019-01-30-year-of-the-dog.md (hunk 4 above)
Comment (indifferentalex) [reply _posts/2019-01-30-year-of-the-dog.md]: If you read it with a pause it sounds bad in my opinion, feels far too forced

Comment (indifferentalex) [reply _posts/2019-01-30-year-of-the-dog.md]: I say go with artistic license and don't put it in

Comment (brainlessdeveloper) [reply _posts/2019-01-30-year-of-the-dog.md]: :man_shrugging: The only case where "however" can go without a comma is when it's a conjunction:

`I love regardless of however much you mess up your punctuation.`

Comment (brainlessdeveloper) [reply _posts/2019-01-30-year-of-the-dog.md]: It's Tom's post and he can do whatever he wants though.

File: _posts/2019-01-30-year-of-the-dog.md [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+This post is a recap of how it has worked for us in 2018.
+
+![store2be developer, early 2018]({{ site.url }}/images/2018_retrospective/grumpy_pallas_cat_c_tambako_the_jaguar_flickr.jpg)
+_store2be developer, early 2018 (photo © Tambako The Jaguar, flickr)_
+
+## Frontend - the React apps
+
+We introduced Typescript in our React apps at the end of 2017, and we are only
+now getting close to completely typed frontend apps. Before Typescript, we had
+suffered from the churn in our JS architecture as we experimented with more
```
[the remark this answers is not in the document]
Comment (brainlessdeveloper) [reply _posts/2019-01-30-year-of-the-dog.md]: The Churn :tm:

File: _posts/2019-01-30-year-of-the-dog.md (hunk 13 above)
Comment (peterfication) [reply _posts/2019-01-30-year-of-the-dog.md]: ups, read the sentence wrong.

File: _posts/2019-01-30-year-of-the-dog.md [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+- Typescript
+- API
+title: What changed in our stack in 2018
+author_staff_member: 03_tom
+---
+
+As our team and codebase grows, we sometimes find the need to cast a look back
+at how things changed, and remember fondly the code we wrote and deleted, the
+abstractions we built and scrapped, how our code used to be, and how much fun
+it was to get it to where it is now.
```
Comment (SoorajChandran) [inline _posts/2019-01-30-year-of-the-dog.md]: I loved the intro <3

File: _posts/2019-01-30-year-of-the-dog.md (hunk 4 above)
Comment (questionmarcus) [reply _posts/2019-01-30-year-of-the-dog.md]: Might go with a wild suggestion: Remove _however_.

We could instead go with
> Despite the numerous changes to the frontend workflow, architectural issues were still making our lives difficult.

The "however" feels a little detached from the previous statements IMO.

File: _posts/2019-01-30-year-of-the-dog.md [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+  deployment smoother as a result. We use
+  [eslint-plugin-graphql](https://github.com/apollographql/eslint-plugin-graphql)
+  to warn us when an app is using a field that has been deprecated.
+
+![store2be developer discovering GraphiQL]({{ site.url }}/images/2018_retrospective/developer_discovering_graphql.jpg)
+_store2be developer discovering GraphiQL (photo © unknown)_
+
+## GraphQL in our Rails API
+
+Adopting GraphQL was in retrospect an easy win in the frontend. It did require
```
Comment (questionmarcus) [inline _posts/2019-01-30-year-of-the-dog.md]:
```suggestion
Adopting GraphQL was, in retrospect, an easy win in the frontend. It did require
```

File: _posts/2019-01-30-year-of-the-dog.md [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+some changes, but towards a more standard workflow with a wide array of tools
+to make our lives easier. It was not so in the API. GraphQL APIs pose a
```
Comment (questionmarcus) [inline _posts/2019-01-30-year-of-the-dog.md]: > it was not so ... in the API.

Feels like it is missing an adjective, y'know.

File: _posts/2019-01-30-year-of-the-dog.md [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+author_staff_member: 03_tom
+---
+
+As our team and codebase grows, we sometimes find the need to cast a look back
+at how things changed, and remember fondly the code we wrote and deleted, the
+abstractions we built and scrapped, how our code used to be, and how much fun
+it was to get it to where it is now.
+
+We are building a platform to make live marketing scalable, measurable and
+targeted — as such we have multiple apps for the various actors of the platform
```
Comment (questionmarcus) [inline _posts/2019-01-30-year-of-the-dog.md]: Nice use of the em-dash 💯 (cheers me up when I see it instead of the hyphen)

File: _posts/2019-01-30-year-of-the-dog.md [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+   logic back to the backend**.
+2. **Data fetching in the frontend is now a lot less complex**, with queries
+   mapping more or less 1-to-1 with views, and mutations to forms and buttons.
+   We are not using any redux-based abstraction for CRUD anymore, and the
+   overall amount of state management code has decreased. State is (or
+   appears) also more localized to where it belongs in the component tree. We
+   found this easier to reason about.
+3. The simplicity, but also the adherence to a well-adopted and documented
+   standard for data fetching made **onboarding easier**. There are a lot of
+   good GraphQL resources for frontend developers. In our experience picking up
```
Comment (questionmarcus) [inline _posts/2019-01-30-year-of-the-dog.md]: Would definitely list a couple of notable examples: even just the graphql.org/learn page was great.

File: _posts/2019-01-30-year-of-the-dog.md [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+In short, the change was received enthusiastically by the whole team once the
+initial hurdles were behind us, and we are very happy with it.
+
+## Backend and infrastructure
+
+We started implementing an event log for our platform, for auditing, debugging
+and in the future decoupling tasks. It opens up a lot of perspectives and
+challenges for 2019.
+
+Another change we have been working at is the migration to Kubernetes. We
```
Comment (questionmarcus) [inline _posts/2019-01-30-year-of-the-dog.md]:
```suggestion
Another change we have been working on is the migration to Kubernetes. We
```

File: _posts/2019-01-30-year-of-the-dog.md (hunk 4 above)
Comment (tomhoule) [reply _posts/2019-01-30-year-of-the-dog.md]: I'll implement your suggestion

File: _posts/2019-01-30-year-of-the-dog.md (hunk 19 above)
Comment (tomhoule) [reply _posts/2019-01-30-year-of-the-dog.md]: :+1:

File: _posts/2019-01-30-year-of-the-dog.md (hunk 20 above)
Comment (tomhoule) [reply _posts/2019-01-30-year-of-the-dog.md]: Replaced with "it was not the same in the API"

File: _posts/2019-01-30-year-of-the-dog.md (hunk 23 above)
Comment (tomhoule) [reply _posts/2019-01-30-year-of-the-dog.md]: :+1:
